src/utils/parser/getScopes.js

```diff
@@ -0,0 +1,280 @@
+// @flow
+
+import { traverseAst } from "./utils/ast";
+import { getSource } from "./sources";
+
+import type { Source, SourceId, Location, SourceScope } from "debugger-html";
+import type { NodePath, Node, Location as BabelLocation } from "babel-traverse";
+
+export type ScopeDeclaration = {
+  type: string,
+  start: Location,
+  end: Location,
+  bindings: any
+};
+
+export type ScopeDeclarations = ScopeDeclaration[];
+
+type GeneratedScope = {
+  start: Location,
+  end: Location,
+  type: string,
+  bindings: any,
+  children: ?(GeneratedScope[])
+};
+
+type TempScope = {
+  type: string,
+  parent: TempScope | null,
+  scopes: Array<TempScope>,
+  loc: BabelLocation,
+  names: any
+};
+
+function _createTempScope(
+  type: string,
+  parent: TempScope | null,
+  loc: BabelLocation
+): TempScope {
+  let result = {
+    type,
+    parent,
+    scopes: [],
+    loc: loc,
+    names: Object.create(null)
+  };
+  if (parent) {
+    parent.scopes.push(result);
+  }
+  return result;
+}
+
+function isNode(node?: Node, type: string): boolean {
+  return node ? node.type === type : false;
+}
+
+function _getFunctionScope(scope: TempScope): TempScope {
+  let s = scope;
+  while (s.type !== "Function" && s.type !== "Script") {
```

Comment:
not something that needs to be changed; but this might be clearer as a recursive function that returns when it terminates. maybe, not sure myself to be honest 😅 i might just have an aversion to while loops

```javascript
function isScopeFunction(scope) { // can be moved out to a helper
  return  (scope.type !== "Function" && scope.type !== "Script")
}

function getFunctionScope(scope) {
  if (!scope.parent || !isScopeFunction(scope)) {
    return scope;
  }
  return getFunctionScope(scope.parent);
}
```